scvi/inference/posterior.py

```diff
@@ -273,176 +286,416 @@ def differential_expression_stats(self, M_sampling=100):
+        :param genes: Subset of genes names
+        :return:
+        Dictionary containing:
+            `scale`
+                Posterior aggregated scale samples of shape (n_samples, n_genes)
+                where n_samples correspond to either:
+                - n_bio_batches * n_cells * n_samples_per_cell
+                or
+                 - n_samples_total
+            `batch`
+                associated batch ids
+
+        """
+        if n_samples is None and n_samples_per_cell is None:
+            n_samples = 500
+        if batchid is None:
+            batchid = np.arange(self.gene_dataset.n_batches)
+
         px_scales = []
+        batch_ids = []
         if selection is None:
             raise ValueError("selections should be a list of cell subsets indices")
         else:
+            selection = np.array(selection)
             if selection.dtype is np.dtype("bool"):
                 selection = np.asarray(np.where(selection)[0].ravel())
         old_loader = self.data_loader
         for i in batchid:
-            idx = np.random.choice(
-                np.arange(len(self.gene_dataset))[selection], n_samples
-            )
-            sampler = SubsetRandomSampler(idx)
-            self.data_loader_kwargs.update({"sampler": sampler})
-            self.data_loader = DataLoader(self.gene_dataset, **self.data_loader_kwargs)
-            px_scales.append(self.get_harmonized_scale(i))
+            # Mode 1: sampling over cells
+            if n_samples is not None:
+                idx = np.random.choice(
+                    np.arange(len(self.gene_dataset))[selection], n_samples
+                )
+                self.update_sampler_indices(idx)
+                scales = self.get_harmonized_scale(i)
+                px_scales.append(scales)
+                batch_ids.append(i * np.ones(len(scales)))
+            elif n_samples_per_cell is not None:
+                self.update_sampler_indices(selection)
+                for _ in range(n_samples_per_cell):
+                    scales = self.get_harmonized_scale(i)
+                    px_scales.append(scales)
+                    batch_ids.append(i * np.ones(len(scales)))
         self.data_loader = old_loader
         px_scales = np.concatenate(px_scales)
-        return px_scales
+        batch_ids = np.concatenate(batch_ids)
 
-    @torch.no_grad()
-    def differential_expression_score(
+        # Filter genes of interest
+        if genes is not None:
+            px_scales = px_scales[:, self.gene_dataset.genes_to_index(genes)]
+
+        return dict(scale=px_scales, batch=batch_ids)
+
+    def get_bayes_factors(
         self,
         idx1: Union[List[bool], np.ndarray],
         idx2: Union[List[bool], np.ndarray],
+        mode: Optional[str] = "vanilla",
         batchid1: Optional[Union[List[int], np.ndarray]] = None,
         batchid2: Optional[Union[List[int], np.ndarray]] = None,
         genes: Optional[Union[List[str], np.ndarray]] = None,
```

Comment:
I'm not so sure about this genes parameter. It doesn't cut down on any computational cost and is an issue for totalVI because we can use this function but not this genes parameter. I think we should remove it and users can filter genes downstream

Reply:
if anything, the genes subsetting should occur in `differential_expression_score`